File: flutter_obdii/lib/views/onboarding_overlay.dart
// Port of OnboardingScreen.kt — intro scrim and bottom-nav highlight.

import 'package:flutter/material.dart';

import '../screenmodels/onboarding_screen_model.dart';

typedef OnboardingCompleteCallback = void Function(bool startDemo);

class OnboardingContentScrim extends StatelessWidget {
  const OnboardingContentScrim({
    super.key,
    required this.pageIndex,
    required this.onPageIndexChange,
    required this.onComplete,
    /// Space reserved above the system inset so the card clears the tab bar.
    required this.bottomInset,
  });

  final int pageIndex;
  final ValueChanged<int> onPageIndexChange;
  final OnboardingCompleteCallback onComplete;
  final double bottomInset;

  @override
  Widget build(BuildContext context) {
    final page = OnboardingScreenModel.pages[pageIndex];
    final compact = OnboardingScreenModel.usesCompactScrim(pageIndex);
    final theme = Theme.of(context);

    return Stack(
      fit: StackFit.expand,
      children: [
        if (!compact)
          ColoredBox(color: Colors.black.withValues(alpha: 0.52))
        else
          Align(
            alignment: Alignment.topCenter,
            child: FractionallySizedBox(
              heightFactor: 0.42,
              widthFactor: 1,
              child: ColoredBox(color: Colors.black.withValues(alpha: 0.2)),
            ),
          ),
        Padding(
          padding: EdgeInsets.fromLTRB(20, 16, 20, bottomInset),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: [
              Align(
                alignment: Alignment.centerRight,
                child: TextButton(
                  onPressed: () => onComplete(false),
                  child: Text(
                    'Skip',
                    style: TextStyle(
                      color: compact
                          ? theme.colorScheme.onSurface
                          : Colors.white,
                    ),
                  ),
                ),
              ),
              Expanded(
                child: Align(
                  alignment: Alignment.bottomCenter,
                  child: LayoutBuilder(
                    builder: (context, constraints) {
                      return ConstrainedBox(
                        constraints: BoxConstraints(
                          maxHeight: constraints.maxHeight,
                        ),
                        child: SingleChildScrollView(
                          child: _OnboardingCard(
                            pageIndex: pageIndex,
                            page: page,
                            theme: theme,
                            onPageIndexChange: onPageIndexChange,
                            onComplete: onComplete,
                          ),
                        ),
                      );
                    },
                  ),
                ),
              ),
            ],
          ),
        ),
      ],
    );
  }
}

class _OnboardingCard extends StatelessWidget {
  const _OnboardingCard({
    required this.pageIndex,
    required this.page,
    required this.theme,
    required this.onPageIndexChange,
    required this.onComplete,
  });

  final int pageIndex;
  final OnboardingPage page;
  final ThemeData theme;
  final ValueChanged<int> onPageIndexChange;
  final OnboardingCompleteCallback onComplete;

  @override
  Widget build(BuildContext context) {
    return Card(
      elevation: 0,
      color: const Color(0xFFF6F7FB),
      shape: RoundedRectangleBorder(
        borderRadius: BorderRadius.circular(12),
        side: BorderSide(
          color: const Color(0xFF172033).withValues(alpha: 0.1),
        ),
      ),
      child: Padding(
        padding: const EdgeInsets.all(20),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            Text(page.title, style: theme.textTheme.headlineSmall),
            const SizedBox(height: 10),
            Text(page.body, style: theme.textTheme.bodyLarge),
            _OnboardingPageHints(pageIndex: pageIndex),
            if (OnboardingScreenModel.showWelcomeSummary(pageIndex)) ...[
              const SizedBox(height: 14),
              const _OnboardingWelcomeSummary(),
            ],
            const SizedBox(height: 16),
            _OnboardingPageIndicators(selectedIndex: pageIndex),
            const SizedBox(height: 16),
            _OnboardingActions(
              pageIndex: pageIndex,
              onPageIndexChange: onPageIndexChange,
              onComplete: onComplete,
            ),
          ],
        ),
      ),
    );
  }
}

class OnboardingNavHighlight extends StatelessWidget {
  const OnboardingNavHighlight({super.key, required this.highlightedIndex});

  final int? highlightedIndex;

  static const _tabCount = 5;

  @override
  Widget build(BuildContext context) {
    if (highlightedIndex == null) return const SizedBox.shrink();

    final highlightColor = Theme.of(context).colorScheme.primary;

    return IgnorePointer(
      child: SizedBox(
        height: 56,
        child: Padding(
          padding: const EdgeInsets.symmetric(horizontal: 4, vertical: 6),
          child: Row(
            children: List.generate(_tabCount, (idx) {
              return Expanded(
                child: Padding(
                  padding: const EdgeInsets.symmetric(horizontal: 2),
                  child: idx == highlightedIndex
                      ? DecoratedBox(
                          decoration: BoxDecoration(
                            color: highlightColor.withValues(alpha: 0.12),
                            borderRadius: BorderRadius.circular(12),
                            border: Border.all(color: highlightColor, width: 2.5),
                          ),
                        )
                      : const SizedBox.shrink(),
                ),
              );
            }),
          ),
        ),
      ),
    );
  }
}

class _OnboardingPageHints extends StatelessWidget {
  const _OnboardingPageHints({required this.pageIndex});

  final int pageIndex;

  @override
  Widget build(BuildContext context) {
    final kind = OnboardingScreenModel.pages[pageIndex].kind;
    switch (kind) {
      case OnboardingPageKind.tabTour:
        if (OnboardingScreenModel.isGaugesDashboardPage(pageIndex)) {
          return const Padding(
            padding: EdgeInsets.only(top: 14),
            child: _OnboardingGaugesLayoutHint(),
          );
        }
        return const SizedBox.shrink();
      case OnboardingPageKind.gaugePicker:
        return const Padding(
          padding: EdgeInsets.only(top: 14),
          child: _OnboardingGaugePickerHint(),
        );
      default:
        return const SizedBox.shrink();
    }
  }
}

class _OnboardingGaugesLayoutHint extends StatelessWidget {
  const _OnboardingGaugesLayoutHint();

  @override
  Widget build(BuildContext context) {
    final labelStyle = Theme.of(context).textTheme.labelLarge?.copyWith(
          fontWeight: FontWeight.w600,
        );

    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: [
        Text('Ring vs list', style: labelStyle),
        const SizedBox(height: 10),
        SegmentedButton<int>(
          segments: const [
            ButtonSegment(value: 0, label: Text('Gauges')),
            ButtonSegment(value: 1, label: Text('List')),
          ],
          selected: const {0},
          onSelectionChanged: (_) {},
        ),
        const SizedBox(height: 10),
        const _OnboardingHintBullet(
          'Gauges shows circular ring tiles; List shows compact rows.',
        ),
        const _OnboardingHintBullet(
          'Drag a gauge on the dashboard to reorder in either view.',
        ),
      ],
    );
  }
}

class _OnboardingGaugePickerHint extends StatelessWidget {
  const _OnboardingGaugePickerHint();

  @override
  Widget build(BuildContext context) {
    final labelStyle = Theme.of(context).textTheme.labelLarge?.copyWith(
          fontWeight: FontWeight.w600,
        );

    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: [
        Text('On this screen', style: labelStyle),
        const SizedBox(height: 8),
        const _OnboardingHintBullet('Use switches to enable or disable gauges.'),
        const _OnboardingHintBullet(
          'Drag rows under Enabled to set dashboard order.',
        ),
        const _OnboardingHintBullet('Search finds any PID in the library.'),
      ],
    );
  }
}

class _OnboardingHintBullet extends StatelessWidget {
  const _OnboardingHintBullet(this.text);

  final String text;

  @override
  Widget build(BuildContext context) {
    return Row(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        const Text('• ', style: TextStyle(fontWeight: FontWeight.bold)),
        Expanded(
          child: Text(text, style: Theme.of(context).textTheme.bodyMedium),
        ),
      ],
    );
  }
}

class _OnboardingWelcomeSummary extends StatelessWidget {
  const _OnboardingWelcomeSummary();

  @override
  Widget build(BuildContext context) {
    final labelStyle = Theme.of(context).textTheme.labelLarge?.copyWith(
          fontWeight: FontWeight.w600,
        );

    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: [
        Text('What you can do', style: labelStyle),
        const SizedBox(height: 8),
        for (final point in OnboardingScreenModel.welcomeSummaryPoints)
          _OnboardingHintBullet(point),
      ],
    );
  }
}

class _OnboardingPageIndicators extends StatelessWidget {
  const _OnboardingPageIndicators({required this.selectedIndex});

  final int selectedIndex;

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);

    return Row(
      mainAxisAlignment: MainAxisAlignment.center,
      children: List.generate(OnboardingScreenModel.pages.length, (index) {
        final selected = index == selectedIndex;
        return Padding(
          padding: const EdgeInsets.symmetric(horizontal: 4),
          child: Container(
            width: selected ? 10 : 8,
            height: selected ? 10 : 8,
            decoration: BoxDecoration(
              shape: BoxShape.circle,
              color: selected
                  ? theme.colorScheme.primary
                  : theme.colorScheme.onSurface.withValues(alpha: 0.25),
            ),
          ),
        );
      }),
    );
  }
}

class _OnboardingActions extends StatelessWidget {
  const _OnboardingActions({
    required this.pageIndex,
    required this.onPageIndexChange,
    required this.onComplete,
  });

  final int pageIndex;
  final ValueChanged<int> onPageIndexChange;
  final OnboardingCompleteCallback onComplete;

  @override
  Widget build(BuildContext context) {
    if (OnboardingScreenModel.isDemoPage(pageIndex)) {
      return Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          FilledButton(
            onPressed: () => onComplete(true),
            child: const Text('Try Demo'),
          ),
          const SizedBox(height: 8),
          TextButton(
            onPressed: () => onComplete(false),
            child: const Text('Get started without Demo'),
          ),
        ],
      );
    }

    if (OnboardingScreenModel.isLastPage(pageIndex)) {
      return const SizedBox.shrink();
    }

    return FilledButton(
      onPressed: () => onPageIndexChange(pageIndex + 1),
      child: const Text('Next'),
    );
  }
}

File: obdii/swiftui/OnboardingOverlay.swift
/**
 * Port of OnboardingScreen.kt / onboarding_overlay.dart — intro scrim and bottom-nav highlight.
 */

import SwiftUI

struct OnboardingContentScrim: View {
  let pageIndex: Int
  let onPageIndexChange: (Int) -> Void
  let onComplete: (Bool) -> Void

  private static let cardBackground = Color(red: 246 / 255, green: 247 / 255, blue: 251 / 255)
  private static let cardBorder = Color(red: 23 / 255, green: 32 / 255, blue: 51 / 255).opacity(0.1)
  private static let navGap: CGFloat = 12

  var body: some View {
    let page = OnboardingScreenModel.pages[pageIndex]
    let compact = OnboardingScreenModel.usesCompactScrim(pageIndex)

    GeometryReader { proxy in
      let bottomInset = proxy.safeAreaInsets.bottom + 49 + Self.navGap

      ZStack(alignment: .bottom) {
        if compact {
          VStack(spacing: 0) {
            Color.black.opacity(0.2)
              .frame(height: proxy.size.height * 0.42)
            Spacer(minLength: 0)
          }
        } else {
          Color.black.opacity(0.52)
        }

        VStack(spacing: 0) {
          HStack {
            Spacer()
            Button("Skip") { onComplete(false) }
              .foregroundStyle(compact ? Color.primary : Color.white)
          }
          .padding(.horizontal, 20)
          .padding(.top, 16)

          Spacer(minLength: 0)

          ScrollView {
            onboardingCard(page: page)
              .padding(.horizontal, 20)
          }
          .frame(maxHeight: max(0, proxy.size.height - bottomInset - 56))
        }
        .padding(.bottom, bottomInset)
      }
      .ignoresSafeArea()
    }
  }

  @ViewBuilder
  private func onboardingCard(page: OnboardingPage) -> some View {
    VStack(alignment: .leading, spacing: 0) {
      Text(page.title)
        .font(.title2)
        .fontWeight(.semibold)

      Text(page.body)
        .font(.body)
        .padding(.top, 10)

      onboardingPageHints

      if OnboardingScreenModel.showWelcomeSummary(pageIndex) {
        onboardingWelcomeSummary
          .padding(.top, 14)
      }

      onboardingPageIndicators
        .padding(.top, 16)

      onboardingActions
        .padding(.top, 16)
    }
    .frame(maxWidth: .infinity, alignment: .leading)
    .padding(20)
    .background(
      RoundedRectangle(cornerRadius: 12)
        .fill(Self.cardBackground)
        .overlay(
          RoundedRectangle(cornerRadius: 12)
            .stroke(Self.cardBorder, lineWidth: 1)
        )
    )
    .padding(.bottom, 8)
  }

  @ViewBuilder
  private var onboardingPageHints: some View {
    switch OnboardingScreenModel.pages[pageIndex].kind {
    case .tabTour:
      if OnboardingScreenModel.isGaugesDashboardPage(pageIndex) {
        onboardingGaugesLayoutHint
          .padding(.top, 14)
      }
    case .gaugePicker:
      onboardingGaugePickerHint
        .padding(.top, 14)
    default:
      EmptyView()
    }
  }

  private var onboardingGaugesLayoutHint: some View {
    VStack(alignment: .leading, spacing: 10) {
      Text("Ring vs list")
        .font(.subheadline)
        .fontWeight(.semibold)

      Picker("Ring vs list", selection: .constant(GaugesDisplayMode.gauges)) {
        ForEach(GaugesDisplayMode.allCases) { mode in
          Text(mode.title).tag(mode)
        }
      }
      .pickerStyle(.segmented)
      .disabled(true)

      OnboardingHintBullet("Gauges shows circular ring tiles; List shows compact rows.")
      OnboardingHintBullet("Drag a gauge on the dashboard to reorder in either view.")
    }
  }

  private var onboardingGaugePickerHint: some View {
    VStack(alignment: .leading, spacing: 8) {
      Text("On this screen")
        .font(.subheadline)
        .fontWeight(.semibold)

      OnboardingHintBullet("Use switches to enable or disable gauges.")
      OnboardingHintBullet("Drag rows under Enabled to set dashboard order.")
      OnboardingHintBullet("Search finds any PID in the library.")
    }
  }

  private var onboardingWelcomeSummary: some View {
    VStack(alignment: .leading, spacing: 8) {
      Text("What you can do")
        .font(.subheadline)
        .fontWeight(.semibold)

      ForEach(OnboardingScreenModel.welcomeSummaryPoints, id: \.self) { point in
        OnboardingHintBullet(point)
      }
    }
  }

  private var onboardingPageIndicators: some View {
    HStack(spacing: 8) {
      ForEach(OnboardingScreenModel.pages.indices, id: \.self) { index in
        let selected = index == pageIndex
        Circle()
          .fill(selected ? Color.accentColor : Color.primary.opacity(0.25))
          .frame(width: selected ? 10 : 8, height: selected ? 10 : 8)
      }
    }
    .frame(maxWidth: .infinity)
  }

  @ViewBuilder
  private var onboardingActions: some View {
    if OnboardingScreenModel.isDemoPage(pageIndex) {
      Button("Try Demo") { onComplete(true) }
        .buttonStyle(.borderedProminent)
        .frame(maxWidth: .infinity)

      Button("Get started without Demo") { onComplete(false) }
        .frame(maxWidth: .infinity)
        .padding(.top, 8)
    } else if !OnboardingScreenModel.isLastPage(pageIndex) {
      Button("Next") { onPageIndexChange(pageIndex + 1) }
        .buttonStyle(.borderedProminent)
        .frame(maxWidth: .infinity)
    }
  }
}

struct OnboardingNavHighlight: View {
  let highlightedIndex: Int?

  var body: some View {
    if let highlightedIndex {
      HStack(spacing: 4) {
        ForEach(0..<5, id: \.self) { index in
          Group {
            if index == highlightedIndex {
              RoundedRectangle(cornerRadius: 12)
                .stroke(Color.accentColor, lineWidth: 2.5)
                .background(
                  RoundedRectangle(cornerRadius: 12)
                    .fill(Color.accentColor.opacity(0.12))
                )
            } else {
              Color.clear
            }
          }
          .frame(maxWidth: .infinity, maxHeight: .infinity)
          .padding(.horizontal, 2)
        }
      }
      .padding(.horizontal, 4)
      .padding(.vertical, 6)
      .frame(height: 49)
      .allowsHitTesting(false)
    }
  }
}

private struct OnboardingHintBullet: View {
  let text: String

  init(_ text: String) {
    self.text = text
  }

  var body: some View {
    HStack(alignment: .top, spacing: 0) {
      Text("• ")
        .fontWeight(.bold)
      Text(text)
        .font(.subheadline)
    }
  }
}
